File: electron/examples/example-agent.py
"""
Example Python agent demonstrating the Local Agent Authorization Protocol

This script shows how to:
1. Request authorization from the Electron app
2. Handle approval/denial
3. Use the access token to make authenticated requests
4. Release the session when done
"""

import requests
import sys
from typing import Optional

# The Electron app listens on this port
ELECTRON_PORT = 49231
BASE_URL = f"http://127.0.0.1:{ELECTRON_PORT}"


def request_authorization(
    agent_name: str,
    description: str
) -> Optional[str]:
    """
    Request authorization from the Electron app
    
    Returns the access token if approved, None if denied
    """
    try:
        response = requests.post(
            f"{BASE_URL}/auth/request",
            json={
                "agentName": agent_name,
                "agentType": "python-script",
                "description": description
            },
            timeout=60  # Give user time to approve
        )
        
        data = response.json()
        
        if data.get("status") == "approved":
            print(f"✓ Authorization approved!")
            print(f"  Token expires at: {data.get('expiresAt')}")
            return data.get("accessToken")
        elif data.get("status") == "denied":
            print(f"✗ Authorization denied: {data.get('reason')}")
            return None
        else:
            print(f"✗ Unexpected response: {data}")
            return None
            
    except requests.exceptions.ConnectionError:
        print(f"✗ Could not connect to Electron app at {BASE_URL}")
        print("  Make sure the Electron app is running")
        return None
    except Exception as e:
        print(f"✗ Authorization request failed: {e}")
        return None


def interact_with_backend(
    token: str,
    data: Optional[dict] = None
) -> Optional[dict]:
    """
    Make a request to the backend through the Electron app's /interact endpoint
    
    The Electron app will:
    1. Validate your agent token
    2. Retrieve the user's access token from cookies
    3. Make the request to the backend with the user's token
    4. Return the backend response to you
    
    The endpoint and method are configured in the Electron app.
    You only need to provide the data payload.
    """
    try:
        response = requests.post(
            f"{BASE_URL}/interact",
            headers={"Authorization": f"Bearer {token}"},
            json=data or {},
            timeout=30
        )
        
        if response.status_code in [401, 403]:
            error_data = response.json()
            print(f"✗ Authorization failed: {error_data.get('message')}")
            print("  Token may have expired or been revoked")
            return None
        
        if response.status_code >= 400:
            print(f"✗ Request failed with status {response.status_code}")
            print(f"  {response.text}")
            return None
        
        return response.json()
        
    except Exception as e:
        print(f"✗ Request failed: {e}")
        return None


def release_session(token: str) -> None:
    """
    Release the authorization session when done
    """
    try:
        headers = {"Authorization": f"Bearer {token}"}
        response = requests.post(
            f"{BASE_URL}/auth/release",
            headers=headers
        )
        
        if response.status_code == 200:
            print("✓ Session released successfully")
        else:
            print(f"⚠ Failed to release session: {response.json()}")
            
    except Exception as e:
        print(f"⚠ Failed to release session: {e}")


def main():
    """
    Example agent workflow
    """
    print("=" * 60)
    print("Example Python Agent - Authorization Demo")
    print("=" * 60)
    print()
    
    # Step 1: Request authorization
    print("Step 1: Requesting authorization...")
    print("  (Check the Electron app for the consent dialog)")
    print()
    
    token = request_authorization(
        agent_name="Example Python Agent",
        description="Demonstrates the Local Agent Authorization Protocol"
    )
    
    if not token:
        print()
        print("Authorization failed. Exiting.")
        sys.exit(1)
    
    print()
    print("-" * 60)
    print()
    
    # Step 2: Use the token to make requests
    print("Step 2: Making authenticated requests...")
    print()
    
    # Example: Check health endpoint (doesn't require auth)
    print("Testing health endpoint (no auth required):")
    try:
        health = requests.get(f"{BASE_URL}/health").json()
        print(f"  Status: {health.get('status')}")
    except Exception as e:
        print(f"  Failed: {e}")
    
    print()
    
    # Example: Try to access backend through /interact
    print("Testing /interact endpoint (backend proxy with user token):")
    print("  The endpoint and method are configured in the Electron app")
    print("  You only send the data payload")
    
    # Example: Send data to backend
    result = interact_with_backend(
        token=token,
        data={
            "query": "example data",
            "filters": {"status": "active"}
        }
    )
    if result:
        print(f"  Response: {result}")
    else:
        print("  Request failed or backend endpoint not available")
    
    print()
    print("-" * 60)
    print()
    
    # Step 3: Release the session
    print("Step 3: Releasing session...")
    release_session(token)
    
    print()
    print("=" * 60)
    print("Demo complete!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: apps/electron/examples/example-agent.js
// Example Node.js agent demonstrating the Local Agent Authorization Protocol

const http = require('http');

const ELECTRON_PORT = 49231;
const BASE_URL = `http://127.0.0.1:${ELECTRON_PORT}`;

/**
 * Request authorization from the Electron app
 */
async function requestAuthorization(agentName, description, requestedScopes) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify({
      agentName,
      agentType: 'nodejs-script',
      description,
      requestedScopes
    });

    const options = {
      hostname: '127.0.0.1',
      port: ELECTRON_PORT,
      path: '/auth/request',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': data.length
      },
      timeout: 60000 // Give user time to approve
    };

    const req = http.request(options, (res) => {
      let body = '';

      res.on('data', (chunk) => {
        body += chunk;
      });

      res.on('end', () => {
        try {
          const response = JSON.parse(body);
          
          if (response.status === 'approved') {
            console.log('✓ Authorization approved!');
            console.log(`  Token expires at: ${new Date(response.expiresAt).toISOString()}`);
            console.log(`  Granted scopes: ${response.scopes.join(', ')}`);
            resolve(response.accessToken);
          } else if (response.status === 'denied') {
            console.log(`✗ Authorization denied: ${response.reason}`);
            resolve(null);
          } else {
            console.log(`✗ Unexpected response:`, response);
            resolve(null);
          }
        } catch (error) {
          reject(error);
        }
      });
    });

    req.on('error', (error) => {
      if (error.code === 'ECONNREFUSED') {
        console.log(`✗ Could not connect to Electron app at ${BASE_URL}`);
        console.log('  Make sure the Electron app is running');
        resolve(null);
      } else {
        reject(error);
      }
    });

    req.on('timeout', () => {
      req.destroy();
      console.log('✗ Authorization request timed out');
      resolve(null);
    });

    req.write(data);
    req.end();
  });
}

/**
 * Make an authenticated request
 */
async function makeAuthenticatedRequest(token, endpoint, method = 'GET') {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: '127.0.0.1',
      port: ELECTRON_PORT,
      path: endpoint,
      method: method,
      headers: {
        'Authorization': `Bearer ${token}`
      }
    };

    const req = http.request(options, (res) => {
      let body = '';

      res.on('data', (chunk) => {
        body += chunk;
      });

      res.on('end', () => {
        try {
          const response = JSON.parse(body);
          
          if (res.statusCode === 401 || res.statusCode === 403) {
            console.log(`✗ Authorization failed: ${response.message}`);
            console.log('  Token may have expired or been revoked');
            resolve(null);
          } else {
            resolve(response);
          }
        } catch (error) {
          reject(error);
        }
      });
    });

    req.on('error', reject);
    req.end();
  });
}

/**
 * Release the authorization session
 */
async function releaseSession(token) {
  return new Promise((resolve) => {
    const options = {
      hostname: '127.0.0.1',
      port: ELECTRON_PORT,
      path: '/auth/release',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`
      }
    };

    const req = http.request(options, (res) => {
      let body = '';

      res.on('data', (chunk) => {
        body += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200) {
          console.log('✓ Session released successfully');
        } else {
          console.log(`⚠ Failed to release session: ${body}`);
        }
        resolve();
      });
    });

    req.on('error', (error) => {
      console.log(`⚠ Failed to release session: ${error.message}`);
      resolve();
    });

    req.end();
  });
}

/**
 * Main example workflow
 */
async function main() {
  console.log('='.repeat(60));
  console.log('Example Node.js Agent - Authorization Demo');
  console.log('='.repeat(60));
  console.log();

  try {
    // Step 1: Request authorization
    console.log('Step 1: Requesting authorization...');
    console.log('  (Check the Electron app for the consent dialog)');
    console.log();

    const token = await requestAuthorization(
      'Example Node.js Agent',
      'Demonstrates the Local Agent Authorization Protocol',
      ['read:data', 'example:scope']
    );

    if (!token) {
      console.log();
      console.log('Authorization failed. Exiting.');
      process.exit(1);
    }

    console.log();
    console.log('-'.repeat(60));
    console.log();

    // Step 2: Use the token
    console.log('Step 2: Making authenticated requests...');
    console.log();

    // Test health endpoint
    console.log('Testing health endpoint (no auth required):');
    const healthReq = http.get(`${BASE_URL}/health`, (res) => {
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => {
        const health = JSON.parse(body);
        console.log(`  Status: ${health.status}`);
        console.log();

        // Continue with protected endpoint test
        continueDemo(token);
      });
    });

    healthReq.on('error', (error) => {
      console.log(`  Failed: ${error.message}`);
      console.log();
      continueDemo(token);
    });

  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

/**
 * Continue demo after health check
 */
async function continueDemo(token) {
  // Test protected endpoint
  console.log('Testing protected endpoint (requires auth):');
  console.log('  Note: This will fail unless you implement the endpoint');
  const result = await makeAuthenticatedRequest(token, '/api/protected');
  if (result) {
    console.log(`  Response:`, result);
  }

  console.log();
  console.log('-'.repeat(60));
  console.log();

  // Step 3: Release session
  console.log('Step 3: Releasing session...');
  await releaseSession(token);

  console.log();
  console.log('='.repeat(60));
  console.log('Demo complete!');
  console.log('='.repeat(60));
}

// Run the example
main();
